### PR TITLE
Fix multiple issues

### LINKED ISSUE #195: Drill files get scaled incorrectly
May be related to #174 ??

* PCB Viewer in KiCad 6.0 shows everything aligned properly.
* JLCPCB.com online gerber viewer shows everything aligned properly.
* tracespace.io/view online gerber viewer shows everything aligned properly.

However, using GerbV will show the drills at the wrong location (wrong scale?).
Here's an ultra-tiny repro:

[GerbV_Bug.zip](https://github.com/gerbv/gerbv/files/12245790/GerbV_Bug.zip)

It appears by removing ALL the commented lines at the top of the drill files, the drills are then properly parsed / shown.

This is why I thought it might be related to #174.  (I'm opening a separate issue until it's determined to be the same.)

### LINKED ISSUE #211: `MIN`/`MAX` macros are broken
The `MIN` and `MAX` macros are written using C++ features.
C11 does not support references.
C11 does not support the `typeof` operator.
However, they are not in the C++-only portion of the header.

Code fails to build when using these macros from C11.

Fixed by #210.

### LINKED ISSUE #212: `clang-format` CI check broken
The recently added CI check for `clang-format` was broken / never tested.

Fixed by #210.

### LINKED ISSUE #213: `drill.c` - Many constraints / invariants are undocumented and/or implied
This reduces maintainability of the code, and increases the likelihood of errors.

This is especially true when new maintainers attempt to fix bugs.

With invariant checking functions implemented, multiple benefits acrue:
1. implicit rules become explicit.
2. if explicit rule fires an assertion, it can more easily be reasoned over (and changed / edited / removed / adjusted)
3. time is not wasted trying to guess if condition was invariant, or if condition was accidental
4. documentation is improved ... and  because it's a live check, the code becomes the documentation

Add functions that will validate the invariants hold true.
E.g., add function to validate invariants for `drill_state_t` 

Fixed by #210

### LINKED ISSUE #214: `drill.c` -- `header_number_format` field should never be `FMT_USER`
Excellon format is clear that only two values
are valid for `header_number_format`, and
which value is used depends on the units:
* Inches ==> FMT_00_0000
* Metric ==> FMT_000_000

Therefore, do not set to FMT_USER on those
occasions, but to the allowed format type.

Also added many more comments to
`drill_parse_header_is_metric()`, explaining
the implied logic.

Fixed by #210.

### LINKED ISSUE #215: `drill.c` -- Detect unhandled conditions when using floating point numbers
See SEI CERT C Coding Standard [FLP32-C](https://wiki.sei.cmu.edu/confluence/display/c/FLP32-C.+Prevent+or+detect+domain+and+range+errors+in+math+functions)

This is fixed by #210

### LINKED ISSUE #216: `drill.c` -- Avoid floating control statements
Braces after control statements (`if`/`else`/`do`/`while`) should be mandatory.

Admittedly, when considered by itself, it would seem that omitting braces seems OK.  However, this increases risk of bugs, especially with macros.  In fact, the old `dprintf()` macro exposed exactly this type of risk:

```C
// Ensure number format is set to FMT_USER (and warn if not true).
if (state->number_format == FMT_USER)
    dprintf("Some debugging information here.")
else
    state->number_format = FMT_USER;
```

The above code has a bug ... it doesn't actually ever set the number format  to `FMT_USER`.

See SEI CERT C Coding Standard [EXP19-C](https://wiki.sei.cmu.edu/confluence/display/c/EXP19-C.+Use+braces+for+the+body+of+an+if%2C+for%2C+or+while+statement).

Fixed with #210.

### LINKED ISSUE #217: `drill.c` -- Limit variable declarations to one per line
See SEI CERT C Coding Standard [DCL04-C](https://wiki.sei.cmu.edu/confluence/display/c/DCL04-C.+Do+not+declare+more+than+one+variable+per+declaration)

Fixed with #210

### LINKED ISSUE #218: `drill.c` -- Use appropriate scope for variables
Reduce scope of variables where reasonable.

See SEI CERT C Coding Standard [DCL19-C](https://wiki.sei.cmu.edu/confluence/display/c/DCL19-C.+Minimize+the+scope+of+variables+and+functions)

Fixed in #210

### LINKED ISSUE #219: `drill.c` -- Mark function parameters `const` where possible
See SEI CERT C Coding Standard [DCL13-C](https://wiki.sei.cmu.edu/confluence/display/c/DCL13-C.+Declare+function+parameters+that+are+pointers+to+values+not+changed+by+the+function+as+const)

At least partially fixed in #210.

### LINKED ISSUE #220: `drill.c` -- unsafe macros are bad
See SEI CERT C Coding Standard [PRE10-C](https://wiki.sei.cmu.edu/confluence/display/c/PRE10-C.+Wrap+multistatement+macros+in+a+do-while+loop) and [PRE00-C](https://wiki.sei.cmu.edu/confluence/display/c/PRE00-C.+Prefer+inline+or+static+functions+to+function-like+macros).

In particular, the `dprintf()` macro was dangerously using `if` without corresponding `else`.  Oh, and it was also redefining a common function (which was defined by the headers in used in the file ... ouch!) ... with different parameters.  Bad, bad, bad....

C11 tried to support variadic macros, but failed to do so fully as passing zero-parameters would fail to compile.   Thus, for `printf()` style macros (variadic), use the dual-parenthesis trick.

Here's an example of the bug that occur with the old macro for `dprintf()`:

```C
// Ensure number format is set to FMT_USER (and warn if not true).
if (state->number_format == FMT_USER)
    dprintf("Some debugging information here.")
else
    state->number_format = FMT_USER;
```

The above code has a bug ... it doesn't actually ever set the number format  to `FMT_USER`.

Fixed with #210.

### LINKED ISSUE #221: `drill.c` -- Pay down accrued technical debt
* See SEI CERT C Coding Standard []()

* Remove unused parameters from internal functions
  * `drill_add_drill_hole()`
  * `new_state()`

* Use consistent order for similar function parameters
  * Reduces mental load
  * Avoids stack reshuffling
  * Easier debugging (stack in same order...)

* Use descriptive variable names

* Reduce cyclomatic complexity of large functions

These are all at least partially addressed by #210

### PR DESCRIPTION
**Pull request by henrygab**
Highlights are as follows:

1. Fix MIN/MAX macros ... they only worked from C++, failed in C11.  See https://github.com/gerbv/gerbv/commit/45bc71acd855787eebb5aa9403baa8be42260266
1. `clang-format` CI check was not applying the checked-in file style.  See https://github.com/gerbv/gerbv/commit/0f74a9ce107174613392dfb7fc04c63cb665fd4f.
1. Convert unsage `dprintf()` to safer `DPRINTF(())`.  See https://github.com/gerbv/gerbv/commit/a831563a667e72d4e87aa7296bbfec0f6b103f5d
1. Fix issue #195.  See https://github.com/gerbv/gerbv/commit/c261beb39de269f311b31da073c50b8f326fed75
1. Add checks for invariants.  See https://github.com/gerbv/gerbv/commit/d3c32dfdbddd1111f10491891e5cf108f36df009 and https://github.com/gerbv/gerbv/commit/f5d9c723f9123002999a20c99c9f1eef7e6a107b

All commits build and pass tests.  See https://github.com/henrygab/gerbv/pull/1 for automated build / test results for each commit.

Fixes #195.
Fixes #211.
Fixes #212.
Fixes #213.
Fixes #214.
Fixes #215.
Fixes #216.
Fixes #217.
Fixes #218.
Fixes #219.
Fixes #220.
Fixes #221.

### CODE REVIEW

**Comment by eyal0** [comment]
It's hard to review with so much going on!  Could this be multiple PRs please?

**Comment by henrygab** [comment]
I broke every significant change into its own commit.

Can you review commit in sequence?

Stop anytime.  These were not undirected commits.  Each was a reduction of work from my initial dev tree.

I'll consider rebase'ing, with changes at any commit.

**Comment by henrygab** [comment]
Or just consider the above 22 separate PRs.  😆

**Comment by henrygab** [comment]
> It's hard to review with so much going on! Could this be multiple PRs please?

To help you with review, I've opened issues to help describe what the changes are addressing.   That's a lot of issues.   Where I could easily do so, I provided references to CERT coding standards.  Thus, if you wonder why it's worth making a given change, the linked web page gives detailed descriptions of the "why" such practices help write solid, bug-free code.

Each commit here is focused on one type of change.   If you find that is not the case, let me know.  I may be open to splitting, rebasing, and force-pushing.

However, given the multi-day timeframe for PR reviews, doing 22 individual PRs just doesn't make sense for either one of us.

**Comment by ooxi** [comment]
@henrygab first of all, thank you very much for your tremendous work! It's very much appreciated!

I have skipped over most commits and they look reasonable. I'll see if GitHub has a way to mark individual commits as reviewed so we can split the effort and merge when all has been reviewed :-)

File: src/drill.c [hunk 1]
```diff
@@ -82,6 +82,16 @@ typedef enum {
     DRILL_DATA
 } drill_file_section_t;
 
+static const char* drill_file_section_list[] = { "NONE", "HEADER", "DATA", NULL };
```
[the remark this answers is not in the document]
Comment (henrygab) [reply src/drill.c]: true.

File: src/drill.c [hunk 2]
```diff
@@ -612,43 +1000,47 @@ parse_drillfile(gerb_file_t* fd, gerbv_HID_Attribute* attr_list, int n_attr, int
...
+                            if (state->autodetect_file_format) {
+                                if (state->number_format != FMT_00_0000) {
                                     /* save metric format definition for later */
                                     state->backup_number_format = state->number_format;
-                                state->number_format = FMT_00_0000;
-                                state->decimals      = 4;
-                                state->unit          = GERBV_UNIT_INCH;
+                                }
+                                state->number_format   = FMT_00_0000;
+                                state->leading_digits  = 4;
```
Comment (henrygab) [inline src/drill.c:1015]: BUG -- This appears to be inverted / swapped.

File: src/drill.c [hunk 3]
```diff
@@ -612,43 +1000,47 @@ parse_drillfile(gerb_file_t* fd, gerbv_HID_Attribute* attr_list, int n_attr, int
...
+                                state->trailing_digits = 2;
```
Comment (henrygab) [inline src/drill.c:1016]: BUG -- This appears to be inverted / swapped.

Comment (henrygab) [reply src/drill.c]: fixed with aac8cdd

File: src/drill.c (hunk 2 above)
Comment (henrygab) [reply src/drill.c]: fixed with aac8cdd

**Comment by henrygab** [comment]
Do you want me to keep my fork / this branch around?  I'm done with this fix, and want to mentally offload this work.

FYI, I'm not really vested in whether this PR gets accepted, heavily modified, or thrown away.  Of course, I **_did_** ensure that each commit in this branch not only builds, but fully passed all the built-in CI tests.  I looked into this because a company I sent a tiny PCB design to hit this bug.

I'm relatively happy to answer questions about any commit; Unfortunately, at some point my memory will fade....

**Comment by eyal0** [comment]
I want to look at it, I've just been swamped.  I'll get to it next week.

**Comment by henrygab** [comment]
I think your rebase resulted in a `clang-format` failure.  I think removing the extra blank line is all that was needed to fix this.

**Comment by eyal0** [comment]
Sorry about that. I've been going through the commits one by one and putting them in as I review them. I rebased in order to see what work remains.

I can avoid the rebasing in the future.

**Comment by henrygab** [comment]
I will defer to your choices in this matter.  At the same time, I cannot ensure this branch is 100% clean (`git bisect` will always have working builds), as requested in prior PRs.  With this understanding, party as you wish on the branch.

**Comment by henrygab** [comment]
FYI, I have created a new PR (PR #226) today.  I re-generated proof of each commit successfully completing the CI checks.
that other PR also fixes a typo in the fix for #195, where the leading and trailing digits were swapped in one case.
Otherwise, all those commits were programatically applied from this branch.

See 100% CI success rate at https://github.com/henrygab/gerbv/pull/3.

  I am therefore closing this PR.

**Pull request closed without merging**